cpp/src/arrow/python/extension_type.h

```diff
@@ -73,5 +73,61 @@ ARROW_PYTHON_EXPORT Status RegisterPyExtensionType(const std::shared_ptr<DataTyp
 
 ARROW_PYTHON_EXPORT Status UnregisterPyExtensionType();
 
+
+// A GenericExtensionType where the extension_name is a variable and thus
+// multiple instances of it can be registered
+
+class ARROW_PYTHON_EXPORT GenericExtensionType : public ExtensionType {
+ public:
+  // GenericExtensionType() : ExtensionType(arrow::null()) {}
+  // GenericExtensionType(std::shared_ptr<DataType> storage_type, std::string extension_name)
+  //  : ExtensionType(storage_type), extension_name_(extension_name) {}
+
+  std::string extension_name() const override { return extension_name_; }
+
+  bool ExtensionEquals(const ExtensionType& other) const override {
+    const auto& other_ext = static_cast<const ExtensionType&>(other);
+    if (other_ext.extension_name() != this->extension_name()) {
+      return false;
+    }
+    if (other_ext.storage_type() != this->storage_type()) {
+      return false;
+    }
+    return true;
```

Comment:
could this call a Python method to compare type parameters? That way the user could override the Python method and check parameters, so that they could register the extension once but have multiple parametrized types. The default would to just return True.